downstream/diarization/expert.py

```diff
@@ -0,0 +1,267 @@
+        return inputs, labels
+
+    # Interface
+    def forward(self, mode, features, labels, lengths, records, **kwargs):
+        """
+        Args:
+            mode: string
+                'train', 'dev' or 'test' for this forward step
+
+            features:
+                list of unpadded features [feat1, feat2, ...]
+                each feat is in torch.FloatTensor and already
+                put in the device assigned by command-line args
+
+            labels:
+                the frame-wise speaker labels
+
+            records:
+                defaultdict(list), by appending contents into records,
+                these contents can be averaged and logged on Tensorboard
+                later by self.log_records every log_step
+
+        Return:
+            loss:
+                the loss to be optimized, should not be detached
+        """
+        lengths = torch.LongTensor(lengths)
+
+        features = pad_sequence(features, batch_first=True)
+        labels = pad_sequence(labels, batch_first=True, padding_value=0).to(features.device)
+        features, labels = self._match_length(features, labels)
+        predicted = self.model(features)
+
+        # cause logits are in (batch, seq, class) and labels are in (batch, seq)
+        # nn.CrossEntropyLoss expect to have (N, class) and (N,) as input
+        # here we flatten logits and labels in order to apply nn.CrossEntropyLoss
+        class_num = predicted.size(-1)
+        loss, perm_idx, perm_list = self.objective(predicted, labels, lengths)
+
+        # get the best label permutation
+        label_perm = get_label_perm(labels, perm_idx, perm_list)
+
+        (correct, num_frames, speech_scored, speech_miss, speech_falarm,
+            speaker_scored, speaker_miss, speaker_falarm,
+            speaker_error) = calc_diarization_error(predicted, label_perm, lengths)
+
+        if speech_scored > 0 and speaker_scored > 0 and num_frames > 0:
+            SAD_MR, SAD_FR, MI, FA, CF, ACC, DER = (
+                speech_miss / speech_scored,
+                speech_falarm / speech_scored,
+                speaker_miss / speaker_scored,
+                speaker_falarm / speaker_scored,
+                speaker_error / speaker_scored,
+                correct / num_frames,
+                (speaker_miss
+                 + speaker_falarm
+                 + speaker_error) / speaker_scored,
+            )
+        else:
+            SAD_MR, SAD_FR, MI, FA, CF, ACC, DER = 0, 0, 0, 0, 0, 0, 0
+        # debug
+        # print("SAD_MR {}, SAD_FR {}, MI {}, FA {}, CF {}, ACC {}, DER {}".format(SAD_MR, SAD_FR, MI, FA, CF, ACC, DER))
+        records['loss'].append(loss.item())
+        records['acc'] += [ACC]
+        records['der'] += [DER]
+
+        return loss
+
+    # interface
+    def log_records(self, mode, records, logger, global_step, batch_ids, total_batch_num, **kwargs):
```

Comment:
Since both train/dev/test loop will call this log_records (differ by the `mode` argument), the `add_scalar` below might mix the numbers from three modes together. Also, since we prefer to use the same tensorboard for all downstream tasks, I will suggest using `f'diarization/{mode}-acc'` and `f'diarization/{mode}-der'` as key names in case of plotting all acc numbers from all the tasks together :)